Smart_Note_Manager.py

```diff
@@ -0,0 +1,200 @@
+
+    def display(self):
+        """
+        Override the display method to show reminder note details.
+        """
+        print("=== Reminder Note ===")
+        super().display()  # Call the parent class display method
+        print(f"Reminder Time: {self.reminder_time.strftime('%Y-%m-%d %H:%M:%S')}")
+
+
+# Step 3: Implement the NotesManager class
+class NotesManager:
+    def __init__(self):
+        """
+        Initialize the notes manager.
+        """
+        self.notes = []  # Store all notes in a list
+        self.next_id = 1  # Assign unique IDs to notes
+
+    def add_note(self, note_type, content, reminder_time=None):
+        """
+        Add a new note to the manager.
+        
+        :param note_type: The type of note ('text' or 'reminder').
+        :param content: The text content of the note.
+        :param reminder_time: The reminder time (only for reminder notes).
+        """
+        if note_type == "text":
+            note = TextNote(content)  # Create a TextNote
+        elif note_type == "reminder":
+            note = ReminderNote(content, reminder_time)  # Create a ReminderNote
+        else:
+            print("Invalid note type. Use 'text' or 'reminder'.")
+            return
+
+        note.id = self.next_id  # Assign a unique ID to the note
+        self.notes.append(note)  # Add the note to the list
+        self.next_id += 1  # Increment the ID counter
+        print(f"Note added successfully with ID: {note.id}")
+
+    def delete_note(self, note_id):
+        """
+        Delete a note by its ID.
+        
+        :param note_id: The ID of the note to delete.
+        """
+        for note in self.notes:
+            if note.id == note_id:
+                self.notes.remove(note)  # Remove the note from the list
+                print(f"Note with ID {note_id} deleted successfully.")
+                return
+        print(f"Note with ID {note_id} not found.")
+
+    def show_notes(self):
+        """
+        Display all stored notes.
+        """
+        if not self.notes:
+            print("No notes found.")
+            return
+
+        for note in self.notes:
+            note.display()  # Call the display method of each note
+            print("-" * 30)
+
+    def search_notes(self, keyword):
+        """
+        Search for notes containing a specific keyword.
+        
+        :param keyword: The keyword to search for.
+        """
+        found_notes = [note for note in self.notes if keyword.lower() in note.content.lower()]
+        if not found_notes:
+            print(f"No notes found containing the keyword '{keyword}'.")
+            return
+
+        print(f"Found {len(found_notes)} note(s) containing the keyword '{keyword}':")
+        for note in found_notes:
+            note.display()
+            print("-" * 30)
+
+
+# The user interface that allows you to interanct with th Notes Manager.
+
+# Step 4: Main Program Execution
+if __name__ == "__main__":
```

Comment:
you should have a main() function that contains the main program execution then call this function under the condition:
```
`if __name__ == "__main__":
   main()
```